### PR TITLE
Let LaTeX obey :confval:`numfig_secnum_depth` for figures, tables, ...

### PR DESCRIPTION
**Pull request by jfbu**
... and code-blocks.

This is new feature and a bit of bugfix.

I I noticed in passing that prior to this PR, there is some non-coherent behaviour of code-block counter compared to figure and table, particularly when docclass is ``'howto'``, and I need to document the change. Documentation not yet added.

The aim is that figure, table, and code-block numbering will be the same in LaTeX as in HTML when ``numfig`` is activated. Thus it is breaking change. Possibly I could add an option to keep LaTeX ignore the ``numfig_secnum_depth`` setting.

It is independent of PR #4166, but motivated by it. Next step will be to get equation numbering obey the new html feature from #4166.

### CODE REVIEW

File: sphinx/texinputs/sphinx.sty
Comment (jfbu) [inline sphinx/texinputs/sphinx.sty]: in fact this branch will be never taken because the writers/latex.py will create sphinx.sty package option nonumfigreset or numfigreset=section, chapter,... I will revise there, I think what I did initially was in case numfig config setting is not set, then no package option is passed to sphinx so this branch is then taken. But I made a last minute change in latex.py so currently this is not the case.

File: sphinx/writers/latex.py
Comment (tk0miya) [inline sphinx/writers/latex.py]: I feel current `__init__()` is too fat. So I propose you to separate this calculation to independent method or function.

**Review by tk0miya — APPROVED** [review]
almost LGTM.
BTW do you have a plan to add testcase for this?
(either is okay to me)

File: sphinx/texinputs/sphinx.sty
Comment (jfbu) [inline sphinx/texinputs/sphinx.sty]: perhaps this avoidance of printing ``0.`` should be applied only at start of document. If for example the numfig reset happens at subsections, this means if in a section a literalblock arises before first subsection, then it drops entirely the prefix numbering, which might be confusing, the number will be typically ``1``, and then the first one in coming subsection will be ``1.1.1``. Perhaps better for former one to be numbered ``1.0.1``

File: sphinx/texinputs/sphinx.sty
Comment (jfbu) [reply sphinx/texinputs/sphinx.sty]: I have modified numfig handling logic in latex.py in second commit, so now indeed this legacy branch will be taken if numfig is False, thus same LaTeX macros as formerly will get executed in that case.

In passing I noticed that with ``'howto'`` the numbering of code-blocks is not coherent with the one of figures and tables, the latter are numbered by default continously, but code-blocks are reset at sections.

File: sphinx/writers/latex.py
Comment (jfbu) [inline sphinx/writers/latex.py]: in this way if numfig=False, neither ``nonumfigreset`` nor ``numfigreset=<foo>`` is used, and the LaTeX code executes exactly as prior to this PR for backwards compatibility.

**Comment by jfbu** [comment]
@tk0miya yes I will add test-case, but it can only check the passing of LaTeX options, as this is only difference in latex output. I will also update docs. In passing I have identified an issue in current Sphinx with code-block numbering in ``'howto'`` case: it does not act like default behaviour of figure and table numbering.

I can now not work immediately on finishing it but will come back later.

File: sphinx/texinputs/sphinx.sty
Comment (jfbu) [reply sphinx/texinputs/sphinx.sty]: I was actually referring to code about 5 lines above that location, sorry.

File: sphinx/texinputs/sphinx.sty
Comment (jfbu) [inline sphinx/texinputs/sphinx.sty]: I will add a check here that the sectioning level determined by ``\spx@opt@numfigreset`` actually exists, using

    \ltx@ifundefined{c@\spx@opt@numfigreset}

like the  ``\ltx@ifundefined{c@chapter}`` used in other branch.

**Comment by tk0miya** [comment]
Okay, I will review this again after updates. Please let me know all your work will be finished.

File: sphinx/writers/latex.py
Comment (jfbu) [inline sphinx/writers/latex.py]: this will need modification once PR #4316 is merged as it modifies self.top_sectionlevel for howto documents

**Comment by jfbu** [comment]
@tk0miya the PR is ready for final review. When #4316 gets merged, I will need to fix conflicts. Also, in last commit I needed to modify some handling of "secnumdepth". By the way the current ``if tocdepth >= SECNUMDEPTH:`` test in latex.py is confusing because SECNUMDEPTH is set to 3, but the howto and manual classes do ``\setcounter{secnumdepth}{2}``, so indeed the test is actually like ``if tocdepth > 2:`` with ``2`` the real default secnumdepth for LaTeX set by the Sphinx doc classes. It thus boil down to the same thing, but I found the current code very confusing, and I have replaced it with something else.

**Review by tk0miya — APPROVED** [review]
LGTM!

**Comment by jfbu** [comment]
Thanks @tk0miya ! I will first merge stable into master and master into this after #4316 is merged into stable, because there will be some merge conflicts. Then I will merge.

**Comment by jfbu** [comment]
Squashed and rebased for cleaner commit history. Will merge once testing completes.

**Pull request merged**
